src/rinoh/util/rkd.py

```diff
@@ -0,0 +1,31 @@
+from collections.abc import MutableMapping
+from weakref import ref
+
+
+
+# http://stackoverflow.com/a/3387975/438249
+class RefKeyDictionary(MutableMapping):
+    """A dictionary that compares keys based on their id (address). Hence, the
+    keys can be mutable."""
+
+    def __init__(self, *args, **kwargs):
+        self.store = dict()
+        self.update(dict(*args, **kwargs))
+
+    def __getitem__(self, obj):
+        obj_weakref, value = self.store[id(obj)]
+        assert obj_weakref() is obj
+        return value
+
+    def __setitem__(self, obj, value):
+        self.store[id(obj)] = ref(obj), value
```

Comment:
You do not keep a hard reference to obj. It can be removed, and after that its id is no longer valid. New created object can take the same id.